src/journal/sd-journal.c

```diff
@@ -1040,20 +1038,23 @@ _public_ int sd_journal_test_cursor(sd_journal *j, const char *cursor) {
         if (r < 0)
                 return r;
 
-        FOREACH_WORD_SEPARATOR(word, l, cursor, ";", state) {
+        for(;;) {
                 _cleanup_free_ char *item = NULL;
-                sd_id128_t id;
                 unsigned long long ll;
+                sd_id128_t id;
                 int k = 0;
 
-                if (l < 2 || word[1] != '=')
-                        return -EINVAL;
+                r = extract_first_word(&cursor, &item, ";", 0);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to parse cursor: %s", cursor);
```

Comment:
Please do not log from this function, it's a library call that is not supposed to log on its own, the caller should. It doesn't log in any other error case, and shouldn't in this one either. Just return r directly...